dubbo-common/src/main/java/com/alibaba/dubbo/common/extension/ExtensionLoader.java

```diff
@@ -602,90 +596,92 @@ private void loadFile(Map<String, Class<?>> extensionClasses, String dir) {
+            if (i > 0) {
+                name = line.substring(0, i).trim();
+                line = line.substring(i + 1).trim();
+            }
+            if (line.isEmpty()) {
+                return;
+            }
+            Class<?> clazz = Class.forName(line, true, classLoader);
+            if (!type.isAssignableFrom(clazz)) {
+                throw new IllegalStateException("Error when load extension class(interface: " +
+                        type + ", class line: " + clazz.getName() + "), class "
+                        + clazz.getName() + "is not subtype of interface.");
+            }
+            if (clazz.isAnnotationPresent(Adaptive.class)) {
+                if (cachedAdaptiveClass == null) {
+                    cachedAdaptiveClass = clazz;
+                } else if (!cachedAdaptiveClass.equals(clazz)) {
+                    throw new IllegalStateException("More than 1 adaptive class found: "
+                            + cachedAdaptiveClass.getClass().getName()
+                            + ", " + clazz.getClass().getName());
+                }
+            } else {
+                try {
+                    clazz.getConstructor(type);
+                    if (cachedWrapperClasses == null) {
+                        cachedWrapperClasses = new ConcurrentHashSet<Class<?>>();
+                    }
+                    cachedWrapperClasses.add(clazz);
+                } catch (NoSuchMethodException e) {
+                    clazz.getConstructor();
+                    if (name == null || name.length() == 0) {
+                        name = findAnnotationName(clazz);
+                        if (name == null || name.length() == 0) {
+                            if (clazz.getSimpleName().length() > type.getSimpleName().length()
+                                    && clazz.getSimpleName().endsWith(type.getSimpleName())) {
+                                name = clazz.getSimpleName().substring(0, clazz.getSimpleName().length() - type.getSimpleName().length()).toLowerCase();
+                            } else {
+                                throw new IllegalStateException("No such extension name for the class " + clazz.getName() + " in the config " + url);
+                            }
+                        }
                     }
-                } // end of while urls
+                    String[] names = NAME_SEPARATOR.split(name);
+                    if (names != null && names.length > 0) {
+                        Activate activate = clazz.getAnnotation(Activate.class);
+                        if (activate != null) {
+                            cachedActivates.put(names[0], activate);
+                        }
+                        for (String n : names) {
+                            cachedNames.putIfAbsent(clazz, n);
+                            if (cachedClasses.get() == null) {
+                                cachedClasses.set(new HashMap<String, Class<?>>());

```

Comment:
At first, i think move `cachedClasses.set()` this will change the original synchronization scope. But outer callers should make sure all accesses to cachedClasses were synchronized, this condition seems never changed.